### ISSUE TITLE
[Feature][DSv4.1] Tracking: every backend's two-level candidate indexer behind CandidateIndexer

### ISSUE THREAD

**Issue by yuan-luo** [sgl-project/sglang#40574]
## Context

DeepSeek-V4.1's low-ratio index layers select attention positions in two levels: the candidate source (layer 20) keeps the best `candidate_topk_blocks = 2048` blocks of 8 compressed positions per query row, the consumers (24 / 28 / 32 / 36) run their top-512 among those blocks. PR [#40352](https://github.com/sgl-project/sglang/pull/40352) put the prefill side of that behind one interface, `CandidateIndexer` (`python/sglang/srt/layers/attention/dsv4/candidate_indexer.py`):

```python
class CandidateIndexer(ABC):
    def publish_prefill(self, inputs: PrefillIndexerInputs, out_positions) -> CandidateMetadata
    def select_prefill(self, published, inputs, out_positions) -> None
    def prefill_tail(self, published, tail_lens) -> CandidateMetadata
```

with two implementations, and the decode side already had one:

| implementation | file | prefill | decode / verify |
|---|---|---|---|
| `DeepGemmCandidateIndexer` (SM100) | `dsv4/candidate_indexer_deep_gemm.py` | block table from the source's dense tiles (`amax8_varlen`, ragged top-k, `sort_candidate_blocks`, DeepGEMM schedule); consumers score only their blocks with `fp8_fp4_paged_sparse_mqa_logits` + `topk_transform_bf16_small` | `publish_decode` / `select_decode`, same kernels over the paged index-K pool |
| `DenseCandidateIndexer` | `dsv4/dense_prefill_indexer.py` | #40217's `dense_prefill_topk`: block ids per request (`PrefillCandidateBlocks`), consumers rebuild a mask per 2 GiB tile and run the plain top-k | — |

`make_candidate_indexer` returns the DeepGEMM indexer on SM100 (with the dense one for prefill under context parallelism) and `None` on everything else. The backend's dense prefill path (`_low_ratio_index_topk_dense`) only calls the three methods; on 4x B200 at 262K context that took the four consumer layers from 41 ms to 2.6 ms each and TTFT from 6.7 s to 5.3 s.

Everything else still scores the whole context and selects through masks inline in `deepseek_v4_backend.py`, or has no DSv4.1 low-ratio path at all. This issue lists what is left, per backend and mode, with the code that has to move.

## Where the candidate indexer runs today

| platform / mode | entry point | how candidates are handled | cost per index layer |
|---|---|---|---|
| SM100 CUDA (and `DeepseekV4TrtllmAttnBackend`, which inherits it), decode + target-verify | `_low_ratio_index_topk_decode` -> `DeepGemmCandidateIndexer.publish_decode` / `select_decode` | block table + DeepGEMM sparse logits; graph variants `candidate_all` / `candidate_c2_all` / `candidate_unfiltered` / `candidate_filtered` (`graph_variants.py`, `decode_cuda_graph_runner.py`) | consumers score 2048 x 8 positions |
| SM100 CUDA, extend (dense fp4 path) | `_low_ratio_index_topk_dense` -> `CandidateIndexer` | block table, consumers sparse | source 43.5 ms, consumer 2.6 ms at 16K rows x 262K |
| SM100 CUDA, extend under CP (`_forward_low_ratio_sources_cp`) | same, but `make_candidate_indexer` hands the DeepGEMM indexer a `DenseCandidateIndexer` for prefill | dense tiles + masks | consumer 41 ms |
| SM100 CUDA, extend inside the prefill CUDA graph | `_low_ratio_index_topk_prefill_graph` | none needed: asserts `width <= candidate_topk_blocks * candidate_block_size`, so every block is a candidate and the plain top-k is the whole selection | — |
| SM90 CUDA, decode + verify | `_low_ratio_index_topk_sm90_decode` | **inline**: Triton `fp4_index_logits_decode` scores all `[B, L]` positions per index layer; the source builds `CandidateMasks(mask=[B, L] bool)` with `select_candidate_blocks`; consumers `masked_fill` + `torch.topk` + `mask_topk_scores` | full context per consumer |
| SM90 CUDA, extend (and SM100 with `SGLANG_DSV41_TORCH_PREFILL_INDEXER=1`) | `_low_ratio_index_topk_torch` | **inline**: per request, per row chunk, bf16 einsum over dequantized K (`get_low_ratio_index_k_dequant`); source publishes `CandidateMasks(request_masks=[[rows_b, lc_b] bool])`, consumers mask + `torch.topk`; the tail cut goes through `_publish_prefill_masks` / `cut_request_masks` | full context per consumer, plus the masks |
| ROCm (`DeepseekV4HipRadixBackend`, selected for `dsv4` on HIP in `attention_registry.py`) | — | no ratio-1/2 (`low_ratio`) path at all: the backend carries the V4 C4 indexer only (`C4IndexerBackendMixin`); its metadata only knows compress ratios 0 / 4 / 128 | DSv4.1 not served |
| NPU (`DeepseekV4AscendAttnBackend`) | — | no DSv4.1 low-ratio path | DSv4.1 not served |

`make_candidate_indexer` and the graph variants are gated on `device_sm >= 100` / `not is_hip()`, so on every non-SM100 target the protocol is simply absent.

### 1. Hopper decode / verify behind the protocol, consumers scoring candidates only

`_low_ratio_index_topk_sm90_decode` (backend ~L3475) is the only decode path outside `CandidateIndexer`. Today each consumer layer scores `[bs, L]` for the whole context and masks 15 of every 16 columns away, and the graph variants of `graph_variants.py` are not built for it.

- Add `publish_decode` / `select_decode` to the `CandidateIndexer` ABC (they exist on `DeepGemmCandidateIndexer` only) and move the SM90 code into an implementation, so the backend's decode dispatch is `self.candidate_indexer.publish_decode(...)` / `select_decode(...)` on every platform.
- Consumer scoring: there is no DeepGEMM paged sparse kernel below SM100. Options: extend `fp4_index_logits_decode` (Triton, `kernels/ops/attention/dsv4/fp4_indexer.py`) with a block-id input so it reads only the 2048 blocks per row (`slots[b, blocks[b, i // 8] * 8 + i % 8]`), or reuse the gathering Triton scorer of #40268 (`candidate_fp4_mqa_logits`, contiguous K) once it takes per-row block ids for a batch.
- Decode graph variants for SM90 once the selection has a fixed shape.

### 2. Torch prefill path behind the protocol

`_low_ratio_index_topk_torch` (backend ~L3555) is Hopper's prefill indexer and the SM100 test oracle. Its publish / consume is interleaved with the per-request row-chunk loop, so it does not fit `publish_prefill` / `select_prefill` as written.

- Restructure into a `TorchCandidateIndexer` (or fold into `DenseCandidateIndexer` with a torch scorer) that publishes `PrefillCandidateBlocks` per request and consumes them by gathering the kept blocks' dequantized K rows (`index_select` over `[2048 * 8]` rows per query row is the natural sparse path for torch) instead of scoring `[rows, lc]` and masking.
- Then `_publish_prefill_masks` / `cut_request_masks` / `CandidateMasks.request_masks` go away, and `enter_late_layer_tail` is the same on every platform.
- Keep the oracle property: the kernel tests (`test_dsv41_prefill_sparse_indexer.py`) compare implementations pairwise, so the torch one should stay the reference the others are checked against.

### 3. DeepGEMM sparse prefill under context parallelism

`make_candidate_indexer` gives the DeepGEMM indexer a `DenseCandidateIndexer` for prefill when `attn_cp_size > 1`, because a CP rank's local rows are interleaved across the prompt and `core.page_table[:num_tokens]` is not their page table. The consumers then pay the dense pass again (41 ms per layer at 262K).

- Build the per-row index page table from `req_to_token` (the way `slot_chunks` in `_low_ratio_index_topk_dense` already derives the slots: `req_to_token[req, j * ratio] // ratio`) so `PrefillIndexerInputs.kv_page_table` is CP-agnostic, then drop `prefill_dense`.
- Needs a CP run to validate; nothing in CI exercises DSv4.1 prefill CP with candidates today.

### 4. ROCm and the other backends

`DeepseekV4HipRadixBackend` (what `attention_registry.py` picks for `dsv4` on HIP) has no ratio-1/2 compressor / indexer path, so DSv4.1 does not run there yet; the Ascend backend has none either. When those land, the candidate indexer should be implemented against the protocol from the start rather than inline: `fp8_fp4_mqa_logits` / the paged sparse kernel are CUDA-only (`_has_dense_fp4_indexer` checks `torch.version.cuda`), so the ROCm implementation would be aiter or Triton (the scorer of item 1 is the natural shared piece).

### 5. SM100 follow-ups inside the DeepGEMM implementation

Not backend coverage, but the remaining cost on the covered path:

- `get_paged_sparse_mqa_logits_metadata` is one 14 ms launch per chunk at 16384 rows (about 290 MB of schedule); all rows of a chunk share one page-table row, so a cheaper schedule for many rows of one request is the next win on the source side.
- The source's own top-512 and the block keys read every score tile twice (`TODO(dark)` in `candidate_indexer_deep_gemm.py`: one kernel for both).
- Verify rows: `publish_decode` pairs rows by `request_ids`; the prefill table uses the same path, but DSpark target-verify with candidates only has the `candidate_filtered` variant.

## Contract for a new implementation

- Prefill: `publish_prefill(inputs, out_positions)` writes the source layer's own top-k (flattened-K columns, `-1` padded, unordered) into `out_positions` and returns the published metadata; `select_prefill(published, inputs, out_positions)` writes a consumer's top-k in the same layout; `prefill_tail(published, tail_lens)` returns the metadata restricted to the last `tail_lens[b]` rows of each request. `PrefillIndexerInputs` carries the fp4 query, head weights, `compress_lens`, `request_starts`, rows and lengths per request, the flattened index K, the index-K pool view and the KV page table; `dense_prefill_indexer.score_tiles(inputs)` yields the dense scores tile by tile under #40217's budget for implementations that need them.
- Decode: `publish_decode(inputs, page_indices, raw_indices)` / `select_decode(published, inputs, page_indices, raw_indices)` over `IndexerInputs` (to be added to the ABC with item 1).
- Equivalence: the published blocks must equal `select_candidate_blocks` on the `-inf`-masked scores block for block; a consumer's selection must agree with `DenseCandidateIndexer.select_prefill` on the same inputs except at rounding-boundary ties (`test/registered/kernels/ops/attention/test_dsv41_prefill_sparse_indexer.py` is parametrizable over implementations); `test_dsv4_nonpaged_indexer.py::TestCandidateIndexerGating` pins what `make_candidate_indexer` returns per platform.

**Comment by CJW-zzz** [comment]
Hi @yuan-luo, I’ve been looking through #40352 and item 3 here. I noticed CP still falls back to `DenseCandidateIndexer`. My understanding is that the interleaved local query rows don’t line up with `core.page_table[:num_tokens]`, so simply enabling the sparse path wouldn’t be correct.

I’d like to work on the `req_to_token`-based mapping you suggested and validate it with a multi-rank CP run. Is anyone already working on this?

I’d like to get started soon. If it’s available, I can start from #40352 and rebase once it lands. Would that work for you?

**Comment by yuan-luo** [comment]
> Hi [@yuan-luo](https://github.com/yuan-luo), I’ve been looking through [#40352](https://github.com/sgl-project/sglang/pull/40352) and item 3 here. I noticed CP still falls back to `DenseCandidateIndexer`. My understanding is that the interleaved local query rows don’t line up with `core.page_table[:num_tokens]`, so simply enabling the sparse path wouldn’t be correct.
> 
> I’d like to work on the `req_to_token`-based mapping you suggested and validate it with a multi-rank CP run. Is anyone already working on this?
> 
> I’d like to get started soon. If it’s available, I can start from [#40352](https://github.com/sgl-project/sglang/pull/40352) and rebase once it lands. Would that work for you?

@WR-zeng Thanks for following up! Glad to see you’re interested in this.

Feel free to start from #40352 and rebase after it lands. I haven't worked on `req_to_token`-based mapping yet, it's fine to move forward.

**Comment by YJR722** [comment]
Hi @yuan-luo, I’d like to take on the SM90 work in items 1 and 2. I’ll start with Hopper prefill (item 2), bringing it behind the CandidateIndexer interface, adding candidate-only scoring, and validating correctness and performance. Then I’ll work on Hopper decode/verify support (item 1).
I’m ready to get started and plan to use [#40352](https://github.com/sgl-project/sglang/pull/40352) as the starting point, rebasing after it merges. Would that work for you? Thanks!

**Comment by yuan-luo** [comment]
> Hi [@yuan-luo](https://github.com/yuan-luo), I’d like to take on the SM90 work in items 1 and 2. I’ll start with Hopper prefill (item 2), bringing it behind the CandidateIndexer interface, adding candidate-only scoring, and validating correctness and performance. Then I’ll work on Hopper decode/verify support (item 1). I’m ready to get started and plan to use [#40352](https://github.com/sgl-project/sglang/pull/40352) as the starting point, rebasing after it merges. Would that work for you? Thanks!

@YJR722 Thanks for following up~ Happy to hear that, just go ahead.

**Comment by CJW-zzz** [comment]
> > Hi [@yuan-luo](https://github.com/yuan-luo), I’ve been looking through [#40352](https://github.com/sgl-project/sglang/pull/40352) and item 3 here. I noticed CP still falls back to `DenseCandidateIndexer`. My understanding is that the interleaved local query rows don’t line up with `core.page_table[:num_tokens]`, so simply enabling the sparse path wouldn’t be correct.
> > I’d like to work on the `req_to_token`-based mapping you suggested and validate it with a multi-rank CP run. Is anyone already working on this?
> > I’d like to get started soon. If it’s available, I can start from [#40352](https://github.com/sgl-project/sglang/pull/40352) and rebase once it lands. Would that work for you?
> 
> [@WR-zeng](https://github.com/WR-zeng) Thanks for following up! Glad to see you’re interested in this.
> 
> Feel free to start from [#40352](https://github.com/sgl-project/sglang/pull/40352) and rebase after it lands. I haven't worked on `req_to_token`-based mapping yet, it's fine to move forward.

Thanks! I’ve started working on this in #40613. Initial single-GPU checks are done, and multi-rank CP validation is still pending. I’ll rebase after #40352 lands. As a student, I’m also learning to keep my B200 rental sessions short 😭

**Comment by CJW-zzz** [comment]
Hi @yuan-luo , I’ve implemented item 3 (DeepGEMM sparse prefill under context parallelism) in #40613, including the CP validation called out in this tracking issue. The PR is ready for review now. Would appreciate it if you could take a look when you have time. Thanks!
